luatest/server_v2.lua

```diff
@@ -0,0 +1,493 @@
+        local pos = 1
+        repeat -- split read string in lines
+            local endpos = string.find(s, '\n', pos)
+            endpos = endpos and endpos - 1 -- strip terminating \n
+            local line = string.sub(s, pos, endpos)
+            if endpos == nil and s ~= '' then
+                -- Line doesn't end with \n or EOF, append it to buffer
+                -- to be checked on next iteration.
+                buf = buf or {}
+                table.insert(buf, line)
+            else
+                if buf ~= nil then
+                    -- Prepend line with buffered data.
+                    table.insert(buf, line)
+                    line = table.concat(buf)
+                    buf = nil
+                end
+                if string.match(line, 'Starting instance') and reset then
+                    found = nil -- server was restarted, reset search
+                else
+                    found = string.match(line, pattern) or found
+                end
+            end
+            pos = endpos and endpos + 2 -- jump to char after \n
+        until pos == nil
+    until s == ''
+    file:close()
+    return found
+end
+
+--
+-- Replication
+--
+
+--- Assert the server follows the source node with the given ID.
+-- Meaning that it replicates from the remote node normally, has already joined
+-- and subscribed.
+--
+-- @number server_id Server ID.
+function Server:assert_follows_upstream(server_id)
+    local status = self:exec(function(id)
+        return box.info.replication[id].upstream.status
+    end, {server_id})
+    local msg = ('%s: server does not follow upstream'):format(self.alias)
+    assertions.assert_equals(status, 'follow', msg)
+end
+
+-- Election
+
+--- Get the election term as seen by the server.
+--
+-- @return number
+function Server:get_election_term()
+    return self:exec(function() return box.info.election.term end)
+end
+
+--- Wait for the server to reach at least the given election term.
+--
+-- @string term Election term to wait for.
+function Server:wait_for_election_term(term)
+    return wait_for_condition('election term', self, self.exec, self, function(t)
+        return box.info.election.term >= t
+    end, {term})
+end
+
+--- Wait for the server to enter the given election state.
+-- Note that if it becomes a leader, it does not mean it is already writable.
+--
+-- @string state Election state to wait for.
+function Server:wait_for_election_state(state)
+    return wait_for_condition('election state', self, self.exec, self, function(s)
+        return box.info.election.state == s
+    end, {state})
+end
+
+--- Wait for the server to become a *writable* election leader.
+function Server:wait_for_election_leader()
+    -- Include read-only property too because if an instance is a leader, it
+    -- does not mean it finished the synchro queue ownership transition. It is
+    -- read-only until that happens. But in tests usually the leader is needed
+    -- as a writable node.
```

Comment:
```suggestion
    -- Include read-only property too because if an instance is a leader, it
    -- does not mean that it has finished the synchro queue ownership transition.
    -- It is read-only until that happens. But in tests usually, the leader is needed
    -- as a writable node.
```

Reply:
Fixed.